File: 300.longest-increasing-subsequence.py
#
# @lc app=leetcode id=300 lang=python3
#
# [300] Longest Increasing Subsequence
#

# @lc code=start
from typing import *
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        max_lengthes = [1] * n
        i = -1
        for i in range(n):
            for j in range(i-1,-1,-1):
                if nums[i] > nums[j] and max_lengthes[i] < 1+ max_lengthes[j]:
                    max_lengthes[i] = 1+ max_lengthes[j]                    
        print(max_lengthes)
        return max(max_lengthes)

        
# @lc code=end



if __name__ == "__main__":
    print(Solution().lengthOfLIS([10,9,2,5,3,7,101,18]))

File: golang/300.longest-increasing-subsequence.go
/*
 * @lc app=leetcode.cn id=300 lang=golang
 *
 * [300] Longest Increasing Subsequence
 */
package main

import "fmt"

// @lc code=start
func lengthOfLIS(nums []int) int {
	dp := []int{1}
	for i := 1; i < len(nums); i++ {
		j := i - 1
		lastMax := 0
		for ; j >= 0; j-- {
			if nums[j] < nums[i] {
				if lastMax < dp[j] {
					lastMax = dp[j]
				}
			}
		}
		dp = append(dp, lastMax+1)
	}
	max := 0
	for i := 0; i < len(dp); i++ {
		if max < dp[i] {
			max = dp[i]
		}
	}
	return max
}

// @lc code=end
func main() {
	fmt.Println(lengthOfLIS([]int{10, 9, 2, 5, 3, 7, 101, 18}))
}
